dbms/src/Interpreters/NullAwareSemiJoinHelper.h

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Logger.h>
+#include <Core/Block.h>
+#include <Interpreters/Join.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+
+namespace DB
+{
+
+enum class NASemiJoinStep : UInt8
+{
+    /// Check other conditions for the right rows whose join key are equal to this left row.
+    /// The join keys of this left row must not have null.
+    CHECK_OTHER_COND,
+    /// Check join key equal condition and other conditions(if any) for the right rows
+    /// in null rows(i.e. all rows in right table with null join key).
+    /// The join keys of this left row must not have null.
+    CHECK_NULL_ROWS_NOT_NULL,
+    /// Like `CHECK_NULL_ROWS_NOT_NULL` except the join keys of this left row must have null.
+    CHECK_NULL_ROWS_NULL,
+    /// Check join key equal condition and other conditions(if any) for all right rows in blocks.
+    /// The join keys of this left row must have null.
+    CHECK_ALL_BLOCKS,
+    /// Work is done.
+    DONE,
+};
+
+enum class NASemiJoinResultType : UInt8
+{
+    FALSE_VALUE,
+    TRUE_VALUE,
+    NULL_VALUE,
+};
+
+template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS>
+class NASemiJoinResult
+{
+public:
+    NASemiJoinResult(size_t row_num, NASemiJoinStep step, const void * map_it);
+
+    /// For convenience, caller can only consider the result of semi join.
+    /// This function will correct the result if it's not semi join.
+    template <NASemiJoinResultType RES>
+    void setResult()
+    {
+        step = NASemiJoinStep::DONE;
+        if constexpr (KIND == ASTTableJoin::Kind::NullAware_LeftSemi)
+        {
+            result = RES;
+            return;
+        }
+        /// For (left) anti semi join
+        if constexpr (RES == NASemiJoinResultType::FALSE_VALUE)
+            result = NASemiJoinResultType::TRUE_VALUE;
+        else if constexpr (RES == NASemiJoinResultType::TRUE_VALUE)
+            result = NASemiJoinResultType::FALSE_VALUE;
+        else
+            result = NASemiJoinResultType::NULL_VALUE;
```

Comment:
It does not need to. But `LeftAnti` also runs this code and it's no harm to return null value because it will be treated as false in the final filter process.